docs/hybridimages.md

```diff
@@ -0,0 +1,241 @@
+in the guest is generally logging its progress to the first serial port.  When
+`create-hybrid-image` runs as part of Jenkins job, verbose logging to the serial
+port means that the Jenkins console log is likely to be helpful in diagnosing
+failures.
+
+If the guest installer performs its own checks and logs the status of those
+checks to the first serial port, `stdout` from QEMU then be used by other tools
+to confirm that the installation completed properly.  For example, in CentOS
+images, the `ks.cfg` `%pre` and `%post` sections are designed to provide the
+status of those phases.  The corresponding `create-image` scripts verify that
+the expected results are found before calling the image creation successful.
+
+OS installers generally do not log their actions to the console very verbosely.
+To get verbose logs, some form a pre-install script (e.g. `%pre` in `ks.cfg` for
+CentOS) is used to tail log files in the background.  When GNU tail is used in a
+Linux guest, this may be done with:
+
+```
+tail -F foo.log bar.log ... >/dev/ttyS0 2>&1 </dev/null &
+```
+
+The `-F` option is used to ask `tail` to watch for logs to come into existence,
+and reopen them when they are rotated.  This allows you to start tailing a log
+before it exists.  The redirection of `stderr` and `stdin` are primarily to make
+it so that `tail` closes all the file descriptors passed to it by its parent
+process.  Without this, some installers will think that the script is intended
+to complete before the pre-installation script finishes and will cause the
+installation to hang forever.
+
+
+### `create-hybrid-image` provides console via VNC
+
+QEMU is started so that the graphical console is available over VNC.  The ip and
+port are shown on the terminal.  Generally, there is no need to connect to the
+graphical console.  However, in the event of a failure it is often helpful to
+connect to the console to poke around inside a VM using a shell.  Most Linux
+distributions have virtual consoles available with `ctrl-alt-F{1,2,3}`.
+
+**WARNING**: The VNC port is accessible via all interfaces.  If the instance has
+an interface on an untrusted network, use of Cloud Firewall to restrict access
+to TCP port 5901 is is highly recommended.
+
+
+### `create-hybrid-image` Private guest network
+
+QEMU networking is configured such that a network that a private network exists
+within the QEMU process.  QEMU intercepts DHCP requests and provides basic
+networking configuration to the guest.  This network has access to other
+networks that are accessible by the zone via a NAT inplementation that exists
+within QEMU.
+
+
+### `create-hybrid-image` arbitrary QEMU arguments
+
+When an image-specific `create-image` script needs to add additional device,
+have specialized control over the boot loader, etc., additional arguments can be
+specified on the command line.
+
+A key example of how this is used is in the CentOS images.  Rather than
```

Comment:
Could we link to the script on github here to illustrate the example, or are we worried that such a link would quickly bitrot?

Reply:
I'll link to the file that will call this once IMAGE-1183 is complete.

Current: https://github.com/joyent/mi-centos-hvm/blob/IMAGE-1183/create-image
Future: https://github.com/joyent/mi-centos-hvm/blob/master/create-image